File: Pkdd.Models/Results/MmpiResult.cs
﻿using Microsoft.EntityFrameworkCore;
using Pkdd.Abstractions.Entity;
using System;
using System.Collections.Generic;
using System.Text;

namespace Pkdd.Models.Results
{
    [Owned]
    public class MmpiResult : IUpdatable<MmpiResult>
    {
        public int Hypochondriasis { get; set; }

        public int Depression { get; set; }

        public int Hysteria { get; set; }

        public int Psychopathia { get; set; }

        public int Masculinity { get; set; }

        public int Paranoia { get; set; }

        public int Psychasthenia { get; set; }

        public int Schizophrenia { get; set; }

        public int Hypomania { get; set; }

        public int Sociality { get; set; }

        public MmpiResult Update(MmpiResult source)
        {
            foreach(var prop in GetType().GetProperties())
            {
                prop.SetValue(this, prop.GetValue(source));
            }
            return this;
        }
    }
}

File: Pkdd/pkdd-ng-client/src/app/models/persons/results/mmpi-result.ts
import { MmpiResultNames, LocaleName } from './mmpi-result-names';

export class MmpiResult {
    constructor(
        public hypochondriasis: number = 0,
        public depression: number = 0,
        public hysteria: number = 0,
        public psychopathia: number = 0,
        public masculinity: number = 0,
        public paranoia: number = 0,
        public psychasthenia: number = 0,
        public schizophrenia: number = 0,
        public hypomania: number = 0,
        public sociality: number = 0,
    ) { }

    public static get keys(): (keyof MmpiResult)[] {
        return Object.keys(new MmpiResult()) as (keyof MmpiResult)[];
    }

    public static toNameValuePairs(mmpiResult: MmpiResult): { name: keyof MmpiResult, value: number }[] {
        return mmpiResult && MmpiResult.keys.map(key => ({ name: key, value: mmpiResult[key] }));
    }

    public static toArray(mmpiResult: MmpiResult): number[] {
        return mmpiResult && MmpiResult.toNameValuePairs(mmpiResult).map(pair => pair.value);
    }

    public static fromArray(values: number[]): MmpiResult {
        const mmpiResult = new MmpiResult();
        MmpiResult.keys.forEach((name, i) => mmpiResult[name] = values[i]);
        return mmpiResult;
    }
}
